### DISCUSSION TITLE
Support for 2FA

### DISCUSSION THREAD

**Discussion by eeldivad** [wg-easy/wg-easy#869] [Ideas]
Is it possible to integrate 2FA solution into this project?  Here's one I found that may be easy to integrate.  Would this be on the roadmap anytime soon?
https://github.com/NHAS/wag

**Comment by peterlewis** [comment]
Thanks for raising. This isn't something that would be likely to feature on the roadmap. My understanding is that it would be unlikely to result in a real world increase in security.

If a private key were to be considered at risk of compromise to the extent that an additional layer of security (e.g. MFA) was considered to be required, then the key pair itself should not be considered safe to use and should be regenerated.

tl;dr - My understanding is that there's no point in having an additional bolt on the door, if the door itself is transparent.

**Reply by offizium-berndstorath** [reply]
noted in #1179
I do not fully understand your point. Why is the door transparent. This definitely would improve security for the ui

**Comment by eeldivad** [comment]
2FA would protect against hackers that are able to steal a single conf file containing our VPN settings.  Most savvy users know enough about security to not get hacked but nothing is 100% and often people get hacked without their knowledge.  To protect the average/novice user, 2FA would be a nice extra layer of protection.  It would also make this project more widely adopted in companies.  I can see small/medium companies looking for free solutions but are afraid to handout these VPN conf files to their users that aren't very tech savvy working from home and possible risks of phishing and higher probability of getting hacked.  Would you be willing to setup a poll to see the amount of interest in 2FA for this project?

**Reply by vithurshanselvarajah** [reply]
To be perfectly fair. 

I do the IT for a few small - medium companies and i haven't handed out a single conf file. the conf file is kept in a folder that the user doesn't have permissions to and wireguard runs as a service allowing it access to it. I rolled this out via a RMM after creating some custom scripts. 

It means even if the user gets infected. They have no admin rights nor permissions to the conf.

**Comment by stanthewizzard** [comment]
coudl be usefull and less strange than defguard (wich implements 2FA)

**Reply by offizium-berndstorath** [reply]
yes but they do it also not only for the ui but also for connecting to it using a special wireguard client which is completely out-of-scope

**Comment by TheRealC0unt** [comment]
Why not encapsulate the web UI in a reverse proxy which could take care of the 2FA? The UI shouldn't be available from the internet anyway, so the enemy is already behind the gates, if the UI would become a suitable target.

**Reply by stanthewizzard** [reply]
that's what I'm doing with authelia

**Reply by stanthewizzard** [reply]
it is
caddy authelia

**Reply by offizium-berndstorath** [reply]
> The UI shouldn't be available from the internet anyway

The UI should be exposable to the internet. The UI should be safe enough that even if its not behind another auth provider (like authelia) it cant be easily penetrated

**Reply by stanthewizzard** [reply]
you can delete / create a client inside the web UI.
Not safe for me

**Reply by offizium-berndstorath** [reply]
> you can delete / create a client inside the web UI. Not safe for me

What does that even mean? Thats to whole purpose of the web ui. Authelia is not inherently more secure then wg-easy. To say I am okay with exposing the UI behing Authelia but not alone makes little to no sense

**Reply by stanthewizzard** [reply]
2fa with authelia

**Comment by richardcrbb** [comment]
Hi my friends, I was configuring my home server from a youtube video which didn´t explain well how to do it, so it said the open-end could be my DNS that I have with duckdns pointing to my public IP address, when I tried to access the UI xxx.duckdns.org:51821 It didn´t allowed me, so I tried many things until I opened the port 51821, set up my devices and everything worked well. But I was thinking I should have 2fa because anyone can access my WG UI, as a second layer of protection. When I read the comment of @TheRealC0unt  I realized I made a mistake, I should¨ve entered to my UI with my local IP 192.168.0.23 let:s say. Now I closed the port because I already have the devices set up. So the KEY @eeldivad is talking about is the QR code scanned from my UI right? as long as nobody has that file I will be okay right?

**Reply by offizium-berndstorath** [reply]
you are correct

**Comment by Urajmal** [comment]
Ok, I see the security risk here from completely different angle. If a user has it on his phone, he is able to export the config and easily send to someone. As Wireguard connections are not limited, the last one connected has a access to server. So IMO 2FA is a very good decision for allowing Wireguard connection.

**Reply by kaaax0815** [reply]
This will not be implemented. This would require a rework of wireguard itself

**Reply by stanthewizzard** [reply]
doable with authelia

**Reply by kaaax0815** [reply]
authelia would only make the ui secure. op was talking about securing wireguard itself

**Reply by stanthewizzard** [reply]
the UI is a good step in my case

**Reply by kaaax0815** [reply]
you are very welcome to add this feature after v15

**Reply by stanthewizzard** [reply]
using it with caddy authelia

**Reply by Urajmal** [reply]
no rush, just a thought, will follow the project!
